### PR TITLE
Springboot kotlin

### PR DESCRIPTION
**Pull request by mohit-sharma-87**
https://github.com/mohit-sharma-87/DevHub-Articles/blob/springboot_kotlin/SpringBoot-Kotlin-MongoDB.md

### CODE REVIEW

File: SpringBoot-Kotlin-MongoDB.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+![Run icon image](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Run_icon_image_bdb6c52ca8.png)
+
+Now load https://localhost:8080/hello on the browser once the build is complete and that will print
+Hello World on your screen.
+
+![Hello World Output](http://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Hello_World_output_279aed0e43.png)
+
+And on cross-validating this from [Postman](https://www.postman.com) we can clearly understand
+that our `Get` API is working perfectly.  
+![Hello World Output](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Postman_output_216cb86538.png)
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:102]: Why is "restaurants" in path in this Postman screenshot? I know it's used later - but it's confusing here.

File: SpringBoot-Kotlin-MongoDB.md [hunk 2]
```diff
@@ -0,0 +1,361 @@
+# Getting Started with backend development in Kotlin using Spring Boot 3 & MongoDB
+
+> This is an introduction article on how to build a RESTful application in Kotlin using
+> [SpringBoot 3](https://spring.io/) and [MongoDB Atlas](https://www.mongodb.com/atlas/database).
+
+## Introduction
+
+I have been a mobile developer for a while and used Kotlin for building apps. But shortly after a
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:8]: In my view - this paragraph isn't needed - just get straight into it "Today we are going...."

File: SpringBoot-Kotlin-MongoDB.md [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+> easy to create stand-alone, production-grade Spring based Applications that you can "just run".
+
+This implies that it's a tool built on the top of Spring framework allowing us to build web
+applications quickly.
+
+`Spring Boot` uses annotations which do the heavy lifting in the background and a few of them we
+have used already like
+
+1. `@SpringBootApplication`
+   This annotation is marked at class level, declare to the code reader(developer) and Spring that
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:119]:
```suggestion
   This annotation is marked at class level, declares to the code reader (developer) and Spring that
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+   it's a Spring Boot project and allows enabling feature which can also be done using
+   `@EnableAutoConfiguration`,`@ComponentScan` & `@Configuration`.
+
+2. `@RequestMapping` & `@RestController`: These annotation provides the routing information.
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:123]:
```suggestion
2. `@RequestMapping` & `@RestController`: This annotation provides the routing information.
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+   Routing is nothing but a mapping of `HTTP` request path (text after `host/`) to classes that
+   have the implementation of these across various `HTTP` method.
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:125]:
```suggestion
   have the implementation of these across various `HTTP` methods.
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+## Introduction
+
+I have been a mobile developer for a while and used Kotlin for building apps. But shortly after a
+brief experience, loved the ease of use and clarity in the language paradigm that gave me
+the confidence to explore more and build more complex systems like backend applications that can
+support Mobile apps.
+
+So today, we are going to build a basic RESTful application that does a little more than a CRUD
+operation and for that we would use
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:14]:
```suggestion
operation and for that we will use
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+
+* `Spring Boot 3` which is one of the popular framework based on Spring, allowing developers to
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:16]:
```suggestion
* `Spring Boot 3` which is one of the popular frameworks based on Spring, allowing developers to
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+  rather than [database schema](https://en.wikipedia.org/wiki/Database_schema).
+
+## Prerequisite
+
+This is a getting started article, so nothing much is needed as a prerequisite.
+But familiarity with [Kotlin](https://kotlinlang.org/) as a programming language, basic
+understanding of [Rest API](https://en.wikipedia.org/wiki/Representational_state_transfer) and [HTTP
+methods](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol) would be helpful.
+
+To help with development activities, we would be
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:28]:
```suggestion
To help with development activities, we will be
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+below.
+
+![Spring initializer](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Spring_initializr_dbbfd7e79a.png)
+
+And since, we want to create REST API with MongoDB as a database let's add the dependency using Add
+Dependency button on the right.
+
+![Spring dependency](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Spring_Depdendency_bb8866de1c.png)
+
+After all the updates our project settings would look like this.
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:53]:
```suggestion
After all the updates our project settings will look like this.
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+   Routing is nothing but a mapping of `HTTP` request path (text after `host/`) to classes that
+   have the implementation of these across various `HTTP` method.
+
+And these annotations are sufficient for building a basic application. Using Spring Boot
+now we will create RESTful web service with all business logic, but we don't have a data container
+which can store or provide data to run these operations.
+
+## Introduction to MongoDB
+
+For our app, we will be using MongoDB as a database which is a document database. It's an
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:133]:
```suggestion
For our app, we will be using MongoDB as the database. It's an
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+open-source, cross-platform and distributed document database which allows building apps with
+flexible schema. This is great as we can focus on building the app rather than defining
+the schema.
+
+We can get started with MongoDB really quickly
+using [MongoDB Atlas](https://www.mongodb.com/atlas/database)
+which is a database on the cloud as a service and has a free forever tier.
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:140]:
```suggestion
which is a database as a service in the cloud and has a free forever tier.
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+
+I would recommend you to explore, the
+[MongoDB Jumpstart series](https://www.youtube.com/watch?v=RGfFpQF0NpE&list=PL4RCxklHWZ9v2lcat4oEVGQhZg6r4IQGV)
+to get familiar with MongoDB and its various services in under 10 minutes.
+
+## Connecting with Spring Boot App & MongoDB
+
+With the basics for MongoDB covered, now let's connect our SpringBoot project with it. Connect
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:148]:
```suggestion
With the basics for MongoDB covered, now let's connect our SpringBoot project to it. Connect
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+    @Field("grade")
+    val rating: String = "",
+    val score: Int = 0
+)
+```
+
+You would have noticed there is nothing fancy about this class except for the annotation. These
+annotations help us to connect or co-relate classes with databases like
+
+* `@Document` declares that this data class represent a document in the Atlas.
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:220]:
```suggestion
* `@Document` declares that this data class represents a document in Atlas.
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+class Controller(@Autowired val repo: Repo) {
+
+    @GetMapping
+    fun getCount(): Int {
+        return repo.findAll().count()
+    }
+}
+```
+
+![model](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Read_72aa6d89de.png)
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:265]: I would point out that this is Postman here again

File: SpringBoot-Kotlin-MongoDB.md [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+``` 
+
+```kotlin
+@GetMapping("/{id}")
+fun getRestaurantById(@PathVariable("id") id: String): Restaurant? {
+    return repo.findByRestaurantId(id)
+}
+```
+
+And again, we would be using [Postman](https://www.postman.com/) to validate the output
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:283]: I would point out here that you're passing the restaurant ID in the call - and say where the ID came from (the sample data set - you picked a random restaurant?

File: SpringBoot-Kotlin-MongoDB.md [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+
+![ready-id](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Read_By_Id_d1752a108b.png)
+
+Let's also validate this against a non-valid `restaurantId` as well.
+
+![ready-null](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Read_Null_637f97f167.png)
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:289]: You should draw a box/ring around "Status" in the Postman screenshot

File: SpringBoot-Kotlin-MongoDB.md [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+```kotlin
+    @GetMapping("/{id}")
+fun getRestaurantById(@PathVariable("id") id: String): ResponseEntity<Restaurant> {
+    val restaurant = repo.findByRestaurantId(id)
+    return if (restaurant != null) ResponseEntity.ok(restaurant) else ResponseEntity
+        .notFound().build()
+}
+```
+
+![ready-404](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Notfound_404_522b7f60ac.png)
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:306]: Again - draw a ring/box around the Status in the postman screenshot

File: SpringBoot-Kotlin-MongoDB.md [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+fun getRestaurantById(@PathVariable("id") id: String): Restaurant? {
+    return repo.findByRestaurantId(id)
+}
+```
+
+And again, we would be using [Postman](https://www.postman.com/) to validate the output
+
+![ready-id](https://mongodb-devhub-cms.s3.us-west-1.amazonaws.com/Read_By_Id_d1752a108b.png)
+
+Let's also validate this against a non-valid `restaurantId` as well.
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:287]: I'd add a comment that your picking an ID that you hope doesn't exits - this time - 4035673

File: SpringBoot-Kotlin-MongoDB.md [hunk 19]
```diff
@@ -0,0 +1,361 @@
...
+fun updateRestaurant(@PathVariable("id") id: String): Restaurant? {
+    return repo.findByRestaurantId(restaurantId = id)?.let {
+        repo.save(it.copy(name = "Update"))
+    }
+}
+```
+
+This is not an ideal way of updating items in the collection as it requires two operations and can
+be improved further if we use the [MongoDB native driver](https://www.mongodb.com/docs/drivers/)
+which allows us to perform complicated operations in minimum steps.
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:338]:
```suggestion
which allows us to perform complicated operations with the minimum of steps.
```

File: SpringBoot-Kotlin-MongoDB.md [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+fun deleteRestaurant(@PathVariable("id") id: String) {
+    repo.findByRestaurantId(id)?.let {
+        repo.delete(it)
+    }
+}
+```
+
+## Summary
+
+Thank you for reading and hopefully find this article informative! The Complete source code of the
```
Comment (ShaneMDB) [inline SpringBoot-Kotlin-MongoDB.md:356]:
```suggestion
Thank you for reading and hopefully you find this article informative! The Complete source code of the
```

**Review by ShaneMDB — CHANGES_REQUESTED** [review]
My comments & some suggestions (couldn't help myself as reading through but to fix some grammer!)

File: SpringBoot-Kotlin-MongoDB.md (hunk 1 above)
Comment (mohit-sharma-87) [reply SpringBoot-Kotlin-MongoDB.md:102]: This is an issue with the screenshot, thanks for pointing it out.

File: SpringBoot-Kotlin-MongoDB.md (hunk 15 above)
Comment (mohit-sharma-87) [reply SpringBoot-Kotlin-MongoDB.md:283]: Added a note

File: SpringBoot-Kotlin-MongoDB.md (hunk 18 above)
Comment (mohit-sharma-87) [reply SpringBoot-Kotlin-MongoDB.md:287]: Language changed.

File: SpringBoot-Kotlin-MongoDB.md (hunk 2 above)
Comment (mohit-sharma-87) [reply SpringBoot-Kotlin-MongoDB.md:8]: Done

**Pull request merged**